File: nativeshell_dart/lib/src/api_model.dart
import 'dart:ui';
import 'util.dart';

extension OffsetExt on Offset {
  Map serialize() => {'x': dx, 'y': dy};

  static Offset deserialize(dynamic position) {
    final map = position as Map;
    return Offset(map['x'], map['y']);
  }

  static Offset? maybeDeserialize(dynamic position) {
    return position != null ? OffsetExt.deserialize(position) : null;
  }
}

extension SizeExt on Size {
  Map serialize() => {'width': width, 'height': height};

  static Size deserialize(dynamic position) {
    final map = position as Map;
    return Size(map['width'], map['height']);
  }

  static Size? maybeDeserialize(dynamic size) {
    return size != null ? SizeExt.deserialize(size) : null;
  }
}

extension RectExt on Rect {
  Map serialize() => {
        'x': left,
        'y': top,
        'width': width,
        'height': height,
      };
  static Rect deserialize(dynamic rect) {
    final map = rect as Map;
    return Rect.fromLTWH(map['x'], map['y'], map['width'], map['height']);
  }

  static Rect? maybeDeserialize(dynamic rect) {
    return rect != null ? RectExt.deserialize(rect) : null;
  }
}

enum GeometryPreference {
  preferFrame,
  preferContent,
}

class Geometry {
  Geometry({
    this.frameOrigin,
    this.frameSize,
    this.contentOrigin,
    this.contentSize,
    this.minFrameSize,
    this.maxFrameSize,
    this.minContentSize,
    this.maxContentSize,
  });

  final Offset? frameOrigin;
  final Size? frameSize;
  final Offset? contentOrigin;
  final Size? contentSize;
  final Size? minFrameSize;
  final Size? maxFrameSize;
  final Size? minContentSize;
  final Size? maxContentSize;

  Geometry copyWith({
    Offset? frameOrigin,
    Size? frameSize,
    Offset? contentOrigin,
    Size? contentSize,
    Size? minFrameSize,
    Size? maxFrameSize,
    Size? minContentSize,
    Size? maxContentSize,
  }) {
    return Geometry(
      frameOrigin: frameOrigin ?? this.frameOrigin,
      frameSize: frameSize ?? this.frameSize,
      contentOrigin: contentOrigin ?? this.contentOrigin,
      contentSize: contentSize ?? this.contentSize,
      minFrameSize: minFrameSize ?? this.minFrameSize,
      maxFrameSize: maxFrameSize ?? this.maxFrameSize,
      minContentSize: minContentSize ?? this.minContentSize,
      maxContentSize: maxContentSize ?? this.maxContentSize,
    );
  }

  Geometry translate(double dx, double dy) {
    return copyWith(
      frameOrigin: frameOrigin?.translate(dx, dy),
      contentOrigin: contentOrigin?.translate(dx, dy),
    );
  }

  Map serialize() => {
        'frameOrigin': frameOrigin?.serialize(),
        'frameSize': frameSize?.serialize(),
        'contentOrigin': contentOrigin?.serialize(),
        'contentSize': contentSize?.serialize(),
        'minFrameSize': minFrameSize?.serialize(),
        'maxFrameSize': maxFrameSize?.serialize(),
        'minContentSize': minContentSize?.serialize(),
        'maxContentSize': maxContentSize?.serialize(),
      };

  static Geometry deserialize(dynamic value) {
    final map = value as Map;
    return Geometry(
      frameOrigin: OffsetExt.maybeDeserialize(map['frameOrigin']),
      frameSize: SizeExt.maybeDeserialize(map['frameSize']),
      contentOrigin: OffsetExt.maybeDeserialize(map['contentOrigin']),
      contentSize: SizeExt.maybeDeserialize(map['contentSize']),
      minFrameSize: SizeExt.maybeDeserialize(map['minFrameSize']),
      maxFrameSize: SizeExt.maybeDeserialize(map['maxFrameSize']),
      minContentSize: SizeExt.maybeDeserialize(map['minContentSize']),
      maxContentSize: SizeExt.maybeDeserialize(map['maxContentSize']),
    );
  }

  @override
  String toString() {
    return serialize().toString();
  }
}

class GeometryFlags {
  GeometryFlags({
    required this.frameOrigin,
    required this.frameSize,
    required this.contentOrigin,
    required this.contentSize,
    required this.minFrameSize,
    required this.maxFrameSize,
    required this.minContentSize,
    required this.maxContentSize,
  });

  final bool frameOrigin;
  final bool frameSize;
  final bool contentOrigin;
  final bool contentSize;
  final bool minFrameSize;
  final bool maxFrameSize;
  final bool minContentSize;
  final bool maxContentSize;

  Map serialize() => {
        'frameOrigin': frameOrigin,
        'frameSize': frameSize,
        'contentOrigin': contentOrigin,
        'contentSize': contentSize,
        'minFrameSize': minFrameSize,
        'maxFrameSize': maxFrameSize,
        'minContentSize': minContentSize,
        'maxContentSize': maxContentSize,
      };

  static GeometryFlags deserialize(dynamic value) {
    final map = value as Map;
    return GeometryFlags(
        frameOrigin: map['frameOrigin'],
        frameSize: map['frameSize'],
        contentOrigin: map['contentOrigin'],
        contentSize: map['contentSize'],
        minFrameSize: map['minFrameSize'],
        maxFrameSize: map['maxFrameSize'],
        minContentSize: map['minContentSize'],
        maxContentSize: map['maxContentSize']);
  }

  @override
  String toString() {
    return serialize().toString();
  }
}

enum WindowFrame {
  /// Normal window frame (includes title and can be resizable).
  regular,

  /// Window frame without title (can be resizable).
  noTitle,

  /// No window frame, can not be resizable.
  noFrame,
}

class WindowStyle {
  WindowStyle({
    this.frame = WindowFrame.regular,
    this.canResize = true,
    this.canClose = true,
    this.canMinimize = true,
    this.canMaximize = true,
    this.canFullScreen = true,
    this.alwaysOnTop = false,
    this.alwaysOnTopLevel,
    this.trafficLightOffset,
  });

  final WindowFrame frame;

  final bool canResize;
  final bool canClose;
  final bool canMinimize;
  final bool canMaximize; // ignored on mac
  final bool canFullScreen;
  final bool alwaysOnTop;

  /// macOS only, corresponds to NSWindowLevel / CGWindowLevel
  /// - 0 - normal window
  /// - 3 - floating window, torn off menu
  /// - 8 - modal panel
  /// - 19 - utility window
  /// - 20 - dock window
  /// - 24 - main menu
  /// - 25 - status window
  /// - 101 - popup menu
  /// - 102 - overlay window
  /// - 200 - help window
  /// - 500 - dragging window
  /// - 1000 - screen saver
  final int? alwaysOnTopLevel;

  /// macOS only and only applicable for WindowFrame.noTitle;
  /// Controls the offset of window traffic light.
  final Offset? trafficLightOffset;

  dynamic serialize() => {
        'frame': enumToString(frame),
        'canResize': canResize,
        'canClose': canClose,
        'canMinimize': canMinimize,
        'canMaximize': canMaximize,
        'canFullScreen': canFullScreen,
        'alwaysOnTop': alwaysOnTop,
        'alwaysOnTopLevel': alwaysOnTopLevel,
        'trafficLightOffset': trafficLightOffset?.serialize(),
      };

  static WindowStyle deserialize(dynamic value) {
    final map = value as Map;
    return WindowStyle(
        frame: enumFromString(
            WindowFrame.values, map['frame'], WindowFrame.regular),
        canResize: map['canResize'],
        canClose: map['canClose'],
        canMinimize: map['canMinimize'],
        canMaximize: map['canMaximize'],
        canFullScreen: map['canFullScreen'],
        alwaysOnTop: map['alwaysOnTop'],
        alwaysOnTopLevel: map['alwaysOnTopLevel'],
        trafficLightOffset:
            OffsetExt.maybeDeserialize(map['trafficLightOffset']));
  }

  @override
  String toString() {
    return serialize().toString();
  }
}

// MacOS specific;
class WindowCollectionBehavior {
  final bool canJoinAllSpaces;
  final bool moveToActiveSpace;
  final bool managed;
  final bool transient;
  final bool stationary;
  final bool participatesInCycle;
  final bool ignoresCycle;
  final bool fullScreenPrimary;
  final bool fullScreenAuxiliary;
  final bool fullScreenNone;
  final bool allowsTiling;
  final bool disallowsTiling;

  WindowCollectionBehavior({
    this.canJoinAllSpaces = false,
    this.moveToActiveSpace = false,
    this.managed = false,
    this.transient = false,
    this.stationary = false,
    this.participatesInCycle = false,
    this.ignoresCycle = false,
    this.fullScreenPrimary = false,
    this.fullScreenAuxiliary = false,
    this.fullScreenNone = false,
    this.allowsTiling = false,
    this.disallowsTiling = false,
  });

  dynamic serialize() => {
        'canJoinAllSpaces': canJoinAllSpaces,
        'moveToActiveSpace': moveToActiveSpace,
        'managed': managed,
        'transient': transient,
        'stationary': stationary,
        'participatesInCycle': participatesInCycle,
        'ignoresCycle': ignoresCycle,
        'fullScreenPrimary': fullScreenPrimary,
        'fullScreenAuxiliary': fullScreenAuxiliary,
        'fullScreenNone': fullScreenNone,
        'allowsTiling': allowsTiling,
        'disallowsTiling': disallowsTiling,
      };
}

enum BoolTransition {
  no,
  noToYes,
  yes,
  yesToNo,
}

class WindowStateFlags {
  bool get maximized =>
      maximizedTransition == BoolTransition.yes ||
      maximizedTransition == BoolTransition.noToYes;

  bool get minimized =>
      minimizedTransition == BoolTransition.yes ||
      minimizedTransition == BoolTransition.noToYes;

  bool get fullScreen =>
      fullScreenTransition == BoolTransition.yes ||
      fullScreenTransition == BoolTransition.noToYes;

  final bool active;

  final BoolTransition maximizedTransition;
  final BoolTransition minimizedTransition;
  final BoolTransition fullScreenTransition;

  WindowStateFlags({
    required this.maximizedTransition,
    required this.minimizedTransition,
    required this.fullScreenTransition,
    required this.active,
  });

  static WindowStateFlags deserialize(dynamic value) {
    final map = value as Map;
    return WindowStateFlags(
        maximizedTransition: enumFromString(
            BoolTransition.values, map['maximized'], BoolTransition.no),
        minimizedTransition: enumFromString(
            BoolTransition.values, map['minimized'], BoolTransition.no),
        fullScreenTransition: enumFromString(
            BoolTransition.values, map['fullScreen'], BoolTransition.no),
        active: map['active']);
  }

  dynamic serialize() => {
        'maximized': enumToString(maximizedTransition),
        'minimized': enumToString(minimizedTransition),
        'fullScreen': enumToString(fullScreenTransition),
        'active': active,
      };

  @override
  String toString() {
    return serialize().toString();
  }
}

File: nativeshell/src/shell/api_model.rs
use std::collections::HashMap;

use crate::codec::Value;

use super::{status_item_manager::StatusItemHandle, HotKeyHandle, MenuHandle, Point, Rect, Size};

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum GeometryPreference {
    PreferFrame,
    PreferContent,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct WindowGeometry {
    pub frame_origin: Option<Point>,
    pub frame_size: Option<Size>,
    pub content_origin: Option<Point>,
    pub content_size: Option<Size>,

    pub min_frame_size: Option<Size>,
    pub max_frame_size: Option<Size>,
    pub min_content_size: Option<Size>,
    pub max_content_size: Option<Size>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WindowGeometryRequest {
    pub geometry: WindowGeometry,
    pub preference: GeometryPreference,
}

impl WindowGeometryRequest {
    // Returns geometry with redundand fields removed (useful when caller
    // supports all fields)
    pub fn filtered_by_preference(self) -> WindowGeometry {
        let mut geometry = self.geometry;

        match self.preference {
            GeometryPreference::PreferFrame => {
                if geometry.frame_origin.is_some() {
                    geometry.content_origin = None;
                }
                if geometry.frame_size.is_some() {
                    geometry.content_size = None;
                }
                if geometry.min_frame_size.is_some() {
                    geometry.min_content_size = None;
                }
                if geometry.max_frame_size.is_some() {
                    geometry.max_content_size = None;
                }
            }
            GeometryPreference::PreferContent => {
                if geometry.content_origin.is_some() {
                    geometry.frame_origin = None;
                }
                if geometry.content_size.is_some() {
                    geometry.frame_size = None;
                }
                if geometry.min_content_size.is_some() {
                    geometry.min_frame_size = None;
                }
                if geometry.max_content_size.is_some() {
                    geometry.max_frame_size = None;
                }
            }
        }

        geometry
    }
}

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WindowActivateRequest {
    pub activate_application: bool,
}

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WindowDeactivateRequest {
    pub deactivate_application: bool,
}

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PopupMenuRequest {
    pub handle: MenuHandle,
    pub position: Point,

    // Windows only, used for menu bar implementation; is specified this
    // rect will keep receiving mouse events
    pub tracking_rect: Option<Rect>,

    // Windows only, menu will not obscure the specified rect
    pub item_rect: Option<Rect>,

    // Windows only, first item will be pre-selected; Use during keyboard navigation
    // in menubar
    pub preselect_first: bool,
}

#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PopupMenuResponse {
    pub item_selected: bool,
}

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct HidePopupMenuRequest {
    pub handle: MenuHandle,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct WindowGeometryFlags {
    pub frame_origin: bool,
    pub frame_size: bool,
    pub content_origin: bool,
    pub content_size: bool,
    pub min_frame_size: bool,
    pub max_frame_size: bool,
    pub min_content_size: bool,
    pub max_content_size: bool,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct DragData {
    pub properties: HashMap<String, Value>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Copy, Clone, PartialEq)]
pub enum DragEffect {
    None,
    Copy,
    Link,
    Move,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DraggingInfo {
    pub location: Point,
    pub data: DragData,
    pub allowed_effects: Vec<DragEffect>,
}
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DragResult {
    pub effect: DragEffect,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ImageData {
    pub width: i32,
    pub height: i32,
    pub bytes_per_row: i32,
    #[serde(with = "serde_bytes")]
    pub data: Vec<u8>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DragRequest {
    pub image: ImageData,
    pub rect: Rect,
    pub allowed_effects: Vec<DragEffect>,
    pub data: DragData,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum WindowFrame {
    Regular,
    NoTitle,
    NoFrame,
}

impl Default for WindowFrame {
    fn default() -> Self {
        WindowFrame::Regular
    }
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct WindowStyle {
    pub frame: WindowFrame,
    pub can_resize: bool,
    pub can_close: bool,
    pub can_minimize: bool,
    pub can_maximize: bool,
    pub can_full_screen: bool,
    pub always_on_top: bool,
    pub always_on_top_level: Option<i64>,
    pub traffic_light_offset: Option<Point>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq)]
pub enum BoolTransition {
    No,
    NoToYes,
    Yes,
    YesToNo,
}

impl Default for BoolTransition {
    fn default() -> Self {
        BoolTransition::No
    }
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct WindowStateFlags {
    pub maximized: BoolTransition,
    pub minimized: BoolTransition,
    pub full_screen: BoolTransition,
    pub active: bool,
}

impl WindowStateFlags {
    pub fn is_minimized(&self) -> bool {
        self.minimized == BoolTransition::Yes || self.minimized == BoolTransition::NoToYes
    }
    pub fn is_maximized(&self) -> bool {
        self.maximized == BoolTransition::Yes || self.maximized == BoolTransition::NoToYes
    }
    pub fn is_full_screen(&self) -> bool {
        self.full_screen == BoolTransition::Yes || self.full_screen == BoolTransition::NoToYes
    }
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct WindowCollectionBehavior {
    pub can_join_all_spaces: bool,
    pub move_to_active_space: bool,
    pub managed: bool,
    pub transient: bool,
    pub stationary: bool,
    pub participates_in_cycle: bool,
    pub ignores_cycle: bool,
    pub full_screen_primary: bool,
    pub full_screen_auxiliary: bool,
    pub full_screen_none: bool,
    pub allows_tiling: bool,
    pub disallows_tiling: bool,
}

//
// Menu
//

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum MenuItemRole {
    About,
    Hide,
    HideOtherApplications,
    ShowAll,
    QuitApplication,
    MinimizeWindow,
    ZoomWindow,
    BringAllToFront,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum MenuRole {
    Window,
    Services,
}

#[derive(serde::Serialize, serde::Deserialize, Default, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Accelerator {
    pub label: String,
    pub alt: bool,
    pub shift: bool,
    pub meta: bool,
    pub control: bool,
}

#[derive(serde::Serialize, serde::Deserialize, PartialEq, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum CheckStatus {
    None,
    CheckOn,
    CheckOff,
    RadioOn,
    RadioOff,
}

impl Default for CheckStatus {
    fn default() -> Self {
        CheckStatus::None
    }
}

#[derive(serde::Serialize, serde::Deserialize, Default, Debug)]
#[serde(rename_all = "camelCase")]
pub struct MenuItem {
    pub id: i64,
    pub title: String,
    pub enabled: bool,
    pub separator: bool,
    pub check_status: CheckStatus,
    pub role: Option<MenuItemRole>,
    pub submenu: Option<MenuHandle>,
    pub accelerator: Option<Accelerator>,
}

impl PartialEq for MenuItem {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

#[derive(serde::Deserialize, Default, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Menu {
    pub role: Option<MenuRole>,
    pub items: Vec<MenuItem>,
}
#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct MenuCreateRequest {
    pub handle: Option<MenuHandle>,
    pub menu: Menu,
}
#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MenuDestroyRequest {
    pub handle: MenuHandle,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MenuAction {
    pub handle: MenuHandle,
    pub id: i64,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MenuOpen {
    pub handle: MenuHandle,
}

#[derive(serde::Serialize, serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct SetMenuRequest {
    pub handle: Option<MenuHandle>,
}

#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Key {
    pub platform: i64,
    pub physical: i64,
    pub logical: Option<i64>,
    pub logical_shift: Option<i64>,
    pub logical_alt: Option<i64>,
    pub logical_alt_shift: Option<i64>,
    pub logical_meta: Option<i64>,
}

#[derive(serde::Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct KeyboardMap {
    pub keys: Vec<Key>,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct HotKeyCreateRequest {
    pub accelerator: Accelerator,
    pub platform_key: i64,
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HotKeyDestroyRequest {
    pub handle: HotKeyHandle,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HotKeyPressed {
    pub handle: HotKeyHandle,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Screen {
    pub id: i64,
    pub frame: Rect,
    pub work_area: Rect,
    pub scaling_factor: f64,
}

//
// StatusItem
//
#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemCreateRequest {}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemDestroyRequest {
    pub handle: StatusItemHandle,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemSetImageRequest {
    pub handle: StatusItemHandle,
    pub image: Vec<ImageData>,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemSetHintRequest {
    pub handle: StatusItemHandle,
    pub hint: String,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemSetHighlightedRequest {
    pub handle: StatusItemHandle,
    pub highlighted: bool,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemShowMenuRequest {
    pub handle: StatusItemHandle,
    pub menu: MenuHandle,
    pub offset: Point,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemGetGeometryRequest {
    pub handle: StatusItemHandle,
}

#[derive(serde::Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemGetScreenIdRequest {
    pub handle: StatusItemHandle,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub enum StatusItemActionType {
    LeftMouseDown,
    LeftMouseUp,
    RightMouseDown,
    RightMouseUp,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusItemAction {
    pub handle: StatusItemHandle,
    pub action: StatusItemActionType,
    pub position: Point,
}
